### PR TITLE
Add priorityMechanisms chain element

### PR DESCRIPTION
**Pull request by glazychev-art**
Signed-off-by: Artem Glazychev <artem.glazychev@xored.com>

## Description
This PR adds `priorityMechanisms` chain element.
The input is a list of mechanisms by priority. On the Request, the chain element sorts the mechanism preferences according to these priorities.

This element will be used on the forwarder.

## Issue link
https://github.com/networkservicemesh/sdk-vpp/issues/638

## How Has This Been Tested?

- [x] Added unit testing to cover
- [ ] Tested manually
- [x] Tested by integration testing
- [ ] Have not tested

## Types of changes

- [ ] Bug fix
- [x] New functionallity
- [ ] Documentation
- [ ] Refactoring
- [ ] CI

### CODE REVIEW

File: pkg/networkservice/common/prioritymechanisms/client.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type priorityMechanismsClient struct {
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/prioritymechanisms/client.go:33]: How do we plan to use this in forwarder-vpp app?

Comment (glazychev-art) [reply pkg/networkservice/common/prioritymechanisms/client.go:33]: We add it to the forwarder's chain, for example after `filtermechanisms` - https://github.com/networkservicemesh/sdk-vpp/blob/main/pkg/networkservice/chains/forwarder/server.go#L152
To pass values, we will use `NSM_MECHANISM_PRIORITY` env that is a prioritized list (`[]string`) - the first value has the highest priority.

For example, if we have
```
            - name: NSM_MECHANISM_PRIORITY
              value: ipsec
```
this means that the mechanismPreferences order will be: `{ IPSec, Wireguard }`

File: pkg/networkservice/common/prioritymechanisms/client.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+		} else {
+			tail = append(tail, mechanism)
+		}
+	}
+	sort.Slice(head, func(i, j int) bool {
+		return priorities[head[i].GetType()] < priorities[head[j].GetType()]
+	})
+
+	return append(head, tail...)
+}
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/prioritymechanisms/client.go:72]: 1. What about a sequence of the same mechanism types? Say like MEMIF, MEMIF with different parameters.

I think we should use `sort.SliceStable`

2. Can we don't create a copy of the mechanism slice here?

File: pkg/networkservice/common/prioritymechanisms/client_test.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+					Type: kernel.MECHANISM,
+				},
+				{
+					Cls:  cls.LOCAL,
+					Type: memif.MECHANISM,
+				},
+			},
+		}
+	}
+	samples := []struct {
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/prioritymechanisms/client_test.go:63]: Consider to cover corner cases for nil and empty cases

File: pkg/networkservice/common/prioritymechanisms/client.go (hunk 1 above)
Comment (denis-tingaikin) [reply pkg/networkservice/common/prioritymechanisms/client.go:33]: Looks good to me.

/cc @edwarnicke

File: pkg/networkservice/common/prioritymechanisms/client.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type priorityMechanismsClient struct {
+	priorities map[string]int
+}
+
+// NewClient - returns a new client chain element that prioritize mechanisms according to the list
+func NewClient(priorities []string) networkservice.NetworkServiceClient {
```
[the remark this answers is not in the document]
Comment (glazychev-art) [reply pkg/networkservice/common/prioritymechanisms/client.go]: Fixed

File: pkg/networkservice/common/prioritymechanisms/client.go (hunk 2 above)
Comment (glazychev-art) [reply pkg/networkservice/common/prioritymechanisms/client.go]: 1. I'm not sure about the same mechanism types. If we have MEMIF and MEMIF, we need additional diff parameters. What if the user needs the second MEMIF? Also, sort.Slice is a bit faster :)

2. Yes, we can. Fixed

File: pkg/networkservice/common/prioritymechanisms/client_test.go (hunk 3 above)
Comment (glazychev-art) [reply pkg/networkservice/common/prioritymechanisms/client_test.go:63]: Fixed

File: pkg/networkservice/common/prioritymechanisms/client.go (hunk 2 above)
Comment (denis-tingaikin) [reply pkg/networkservice/common/prioritymechanisms/client.go]: >  If we have MEMIF and MEMIF, we need additional diff parameters. What if the user needs the second MEMIF? Also, sort.Slice is a bit faster :)

Note: We shouldn't change the order of MEMIF and MEMIF sequence. So we need to use a stable sort.

Comment (glazychev-art) [reply pkg/networkservice/common/prioritymechanisms/client.go]: Fixed

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**